source/common/http/filter/squash_filter.cc

```diff
@@ -0,0 +1,340 @@
+  the motivation on why this is needed.
+*/
+std::string SquashFilterConfig::replaceEnv(const std::string& attachment_template) {
+  std::string s;
+
+  auto end_last_match = attachment_template.begin();
+
+  auto replaceEnvVarInTemplateCallback =
+      [&s, &attachment_template,
+       &end_last_match](const std::match_results<std::string::const_iterator>& match) {
+        auto start_match = attachment_template.begin() + match.position(0);
+
+        s.append(end_last_match, start_match);
+
+        std::string envar_name = match[1].str();
+        const char* envar_value = std::getenv(envar_name.c_str());
+        if (envar_value == nullptr) {
+          ENVOY_LOG(warn, "Squash: no environment variable named {}.", envar_name);
+        } else {
+          s.append(envar_value);
+        }
+        end_last_match = start_match + match.length(0);
+      };
+
+  std::sregex_iterator begin(attachment_template.begin(), attachment_template.end(), ENV_REGEX),
+      end;
+  std::for_each(begin, end, replaceEnvVarInTemplateCallback);
+  s.append(end_last_match, attachment_template.end());
+
+  return s;
+}
+
+SquashFilter::SquashFilter(SquashFilterConfigSharedPtr config, Upstream::ClusterManager& cm)
+    : config_(config), state_(State::Initial), debugAttachmentPath_(), delay_timer_(nullptr),
+      attachment_timeout_timer_(nullptr), in_flight_request_(nullptr), cm_(cm),
+      decoder_callbacks_(nullptr) {}
+
+SquashFilter::~SquashFilter() {}
+
+void SquashFilter::onDestroy() { cleanup(); }
+
+FilterHeadersStatus SquashFilter::decodeHeaders(HeaderMap& headers, bool) {
+
+  // check for squash header
+  if (!headers.get(Headers::get().XSquashDebug)) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  ENVOY_LOG(info, "Squash: Holding request and requesting debug attachment");
+
+  MessagePtr request(new RequestMessageImpl());
+  request->headers().insertContentType().value().setReference(
+      Headers::get().ContentTypeValues.Json);
+  request->headers().insertPath().value().setReference(POST_ATTACHMENT_PATH);
+  request->headers().insertHost().value().setReference(SERVER_AUTHORITY);
+  request->headers().insertMethod().value().setReference(Headers::get().MethodValues.Post);
+  request->body().reset(new Buffer::OwnedImpl(config_->attachmentJson()));
+
+  state_ = State::CreateConfig;
+  in_flight_request_ = cm_.httpAsyncClientForCluster(config_->clusterName())
+                           .send(std::move(request), *this, config_->requestTimeout());
+
+  if (in_flight_request_ == nullptr) {
+    ENVOY_LOG(info, "Squash: can't create request for squash server");
+    state_ = State::Initial;
+    return FilterHeadersStatus::Continue;
+  }
+
+  attachment_timeout_timer_ =
+      decoder_callbacks_->dispatcher().createTimer([this]() -> void { doneSquashing(); });
+  attachment_timeout_timer_->enableTimer(config_->attachmentTimeout());
+  // check if the timer expired inline.
+  if (state_ == State::Initial) {
```

Comment:
Do you think it should be valid config to have a 0s timeout?

Reply:
Ideally not. My concern was that if the timer does expire inline and I don't handle this case, the filter iteration will not continue, and the request will be lost.
I am open to suggestions!